File: frontend/ios/App/App/SecureSessionPlugin.swift
import Capacitor
import Foundation
import Security

@objc(SecureSessionPlugin)
public class SecureSessionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SecureSessionPlugin"
    public let jsName = "SecureSession"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise)
    ]

    private let service = "com.fridgeboard.app.secure-session"

    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let data = read(key: key) else {
            return call.resolve(["value": NSNull()])
        }
        call.resolve(["value": String(data: data, encoding: .utf8) as Any])
    }

    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else {
            return call.reject("key and value are required")
        }
        guard let data = value.data(using: .utf8), write(key: key, data: data) else {
            return call.reject("secure storage write failed")
        }
        call.resolve()
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            return call.reject("key is required")
        }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
        call.resolve()
    }

    private func baseQuery(key: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
    }

    private func read(key: String) -> Data? {
        var query = baseQuery(key: key)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: AnyObject?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess else { return nil }
        return result as? Data
    }

    private func write(key: String, data: Data) -> Bool {
        let query = baseQuery(key: key)
        let attributes: [String: Any] = [
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if status == errSecItemNotFound {
            var insert = query
            insert.merge(attributes) { _, new in new }
            return SecItemAdd(insert as CFDictionary, nil) == errSecSuccess
        }
        return status == errSecSuccess
    }
}

File: frontend/android/app/src/main/java/com/fridgeboard/app/SecureSessionPlugin.java
package com.fridgeboard.app;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import android.content.Context;

@CapacitorPlugin(name = "SecureSession")
public class SecureSessionPlugin extends Plugin {
    private SecureSessionStore store() {
        return new SecureSessionStore(getBridge().getContext());
    }

    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");
        if (key == null || key.isEmpty()) {
            call.reject("key is required");
            return;
        }
        try {
            String value = store().get(key);
            JSObject result = new JSObject();
            result.put("value", value);
            call.resolve(result);
        } catch (Exception exception) {
            String code = SecureSessionStore.isRecoverableKeyFailure(exception)
                    ? "SECURE_STORAGE_KEY_MISMATCH"
                    : "SECURE_STORAGE_READ_FAILED";
            call.reject("secure storage read failed", code, exception);
        }
    }

    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        if (key == null || key.isEmpty() || value == null) {
            call.reject("key and value are required");
            return;
        }
        try {
            store().set(key, value);
            call.resolve();
        } catch (Exception exception) {
            String code = SecureSessionStore.isRecoverableKeyFailure(exception)
                    ? "SECURE_STORAGE_KEY_MISMATCH"
                    : "SECURE_STORAGE_WRITE_FAILED";
            call.reject("secure storage write failed", code, exception);
        }
    }

    @PluginMethod
    public void reset(PluginCall call) {
        try {
            resetStorage(getBridge().getContext());
            call.resolve();
        } catch (Exception exception) {
            call.reject("secure storage reset failed", "SECURE_STORAGE_RESET_FAILED", exception);
        }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");
        if (key == null || key.isEmpty()) {
            call.reject("key is required");
            return;
        }
        try {
            store().remove(key);
            call.resolve();
        } catch (Exception exception) {
            call.reject("secure storage remove failed", "SECURE_STORAGE_REMOVE_FAILED", exception);
        }
    }

    private void resetStorage(Context context) throws Exception {
        new SecureSessionStore(context).reset();
    }
}
